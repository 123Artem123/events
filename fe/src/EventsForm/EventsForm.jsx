import React, { useState } from "react";
import TextField from "@material-ui/core/TextField";
import Button from "@material-ui/core/Button";
import Grid from "@material-ui/core/Grid";
import { DatePicker } from "@material-ui/pickers";
import PropTypes from "prop-types";
import { makeStyles } from "@material-ui/core/styles";
import { useTheme } from "@material-ui/core/styles";
import useMediaQuery from "@material-ui/core/useMediaQuery";

const useStyles = makeStyles({
  input: {
    marginBottom: "1rem",
  },
});

const EMAIL_REGEXP = /\S+@\S+\.\S+/;
const FIELD_REGEXP = /^\S{1,64}$/;

const EventsForm = ({ onSubmit }) => {
  const theme = useTheme();
  const matches = useMediaQuery(theme.breakpoints.up("sm"));
  const { input } = useStyles();
  const [email, setEmail] = useState("");
  const [emailError, setEmailError] = useState(null);
  const [fn, setFn] = useState("");
  const [fnError, setFnError] = useState(null);
  const [ln, setLn] = useState("");
  const [lnError, setLnError] = useState(null);
  const [date, setDate] = useState(new Date());

  const validate = (regExp, value) => !regExp.test(value);

  const onEmailChange = (event) => {
    const value = event.target.value;
    setEmail(value);
    setEmailError(validate(EMAIL_REGEXP, value));
  };

  const onFnChange = (event) => {
    const value = event.target.value;
    setFn(value);
    setFnError(validate(FIELD_REGEXP, value));
  };

  const onLnChange = (event) => {
    const value = event.target.value;
    setLn(value);
    setLnError(validate(FIELD_REGEXP, value));
  };

  const onReset = () => {
    setEmail("");
    setLn("");
    setFn("");
    setDate(new Date());
  };

  const onFormSubmit = (event) => {
    event.preventDefault();
    const emailErr = validate(EMAIL_REGEXP, email);
    setEmailError(emailErr);
    const fnErr = validate(FIELD_REGEXP, fn);
    setFnError(fnErr);
    const lnErr = validate(FIELD_REGEXP, ln);
    setLnError(lnErr);
    if (!emailErr && !fnErr && !lnErr) {
      onSubmit({
        email,
        date: date.toISOString().substring(0, 10),
        firstName: fn,
        lastName: ln,
      });
    }
  };

  return (
    <form noValidate autoComplete="off" onSubmit={onFormSubmit}>
      <Grid
        container
        direction="column"
        justifyContent="center"
        alignItems="stretch"
      >
        <TextField
          inputProps={{ "data-testid": "EventsForm-email" }}
          error={emailError}
          onBlur={(e) => setEmailError(validate(EMAIL_REGEXP, e?.target.value))}
          value={email}
          onChange={onEmailChange}
          className={input}
          label="Email"
          variant="outlined"
        />
        <TextField
          inputProps={{ "data-testid": "EventsForm-fn" }}
          error={fnError}
          onBlur={(e) => setFnError(validate(FIELD_REGEXP, e?.target.value))}
          value={fn}
          onChange={onFnChange}
          className={input}
          label="First Name"
          variant="outlined"
        />
        <TextField
          inputProps={{ "data-testid": "EventsForm-ln" }}
          error={lnError}
          onBlur={(e) => setLnError(validate(FIELD_REGEXP, e?.target.value))}
          value={ln}
          onChange={onLnChange}
          className={input}
          label="Last Name"
          variant="outlined"
        />
        <DatePicker
          disablePast
          className={input}
          label="Date"
          value={date}
          onChange={setDate}
          inputVariant="outlined"
        />
        <Grid
          container
          direction={matches ? "row" : "column"}
          justifyContent="space-evenly"
          alignItems="stretch"
          spacing={2}
        >
          <Grid item>
            <Button
              onClick={onReset}
              size="large"
              type="reset"
              color="secondary"
              variant="outlined"
              fullWidth
            >
              Clear
            </Button>
          </Grid>
          <Grid item>
            <Button
              data-testid="EventsForm-submit"
              disabled={emailError || fnError || lnError}
              size="large"
              type="submit"
              color="primary"
              variant="outlined"
              fullWidth
            >
              Submit
            </Button>
          </Grid>
        </Grid>
      </Grid>
    </form>
  );
};

EventsForm.propTypes = {
  onSubmit: PropTypes.func.isRequired,
};

export default EventsForm;
