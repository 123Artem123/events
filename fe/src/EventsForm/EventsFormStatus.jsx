import React from "react";
import Box from "@material-ui/core/Box";
import CheckCircleIcon from "@material-ui/icons/CheckCircle";
import CancelIcon from "@material-ui/icons/Cancel";
import Button from "@material-ui/core/Button";
import Typography from "@material-ui/core/Typography";
import PropTypes from "prop-types";
import { makeStyles } from "@material-ui/core/styles";

const useStyles = makeStyles({
    marginTop: {
      marginTop: "1rem",
    },
  });

const EventsFormStatus = ({ onBack, success, error }) => {
  const { marginTop } = useStyles();

  return (
    <Box
      display="flex"
      justifyContent="center"
      flexDirection="column"
      alignItems="center"
    >
      {success && (
        <CheckCircleIcon
          className={marginTop}
          fontSize="large"
          color="primary"
        />
      )}
      <Typography variant="h6" align="center">
        {success && "Your event has been successfully created!"}
        {error?.message}
      </Typography>

      {error?.details?.map((message) => (
        <Box display="flex" justifyContent="center" alignItems="center" className={marginTop}>
          <CancelIcon fontSize="small" color="error" />
          <Typography key={message}>{message}</Typography>
        </Box>
      ))}
      <Button
        className={marginTop}
        onClick={onBack}
        color="primary"
        variant="outlined"
        size="large"
      >
        Go Back
      </Button>
    </Box>
  );
};

EventsFormStatus.propTypes = {
  onBack: PropTypes.func.isRequired,
  success: PropTypes.bool.isRequired,
};

export default EventsFormStatus;
