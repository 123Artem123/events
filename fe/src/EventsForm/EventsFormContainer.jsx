import React, { useState } from "react";
import Paper from "@material-ui/core/Paper";
import Typography from "@material-ui/core/Typography";
import Box from "@material-ui/core/Box";
import CircularProgress from "@material-ui/core/CircularProgress";
import { makeStyles, createStyles } from "@material-ui/core/styles";
import { useTheme } from "@material-ui/core/styles";
import useMediaQuery from "@material-ui/core/useMediaQuery";
import EventsForm from "./EventsForm";
import EventsFormStatus from "./EventsFormStatus";

const useStyles = makeStyles((theme) =>
  createStyles({
    loadIndicator: {
      padding: "2rem",
    },
    container: {
      padding: "2rem",
      width: (matches) =>
        matches ? `${theme.breakpoints.values.sm.valueOf()}px` : "100%",
    },
  })
);

const EventsFormContainer = () => {
  const theme = useTheme();
  const matches = useMediaQuery(theme.breakpoints.up("sm"));
  const { loadIndicator, container } = useStyles(matches);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState(null);
  const [success, setSuccess] = useState(false);

  const onSubmit = async (event) => {
    try {
      setLoading(true);
      const rawResult = await fetch(process.env.REACT_APP_API_URL, {
        method: "POST",
        headers: {
          Accept: "application/json",
          "Content-Type": "application/json",
        },
        body: JSON.stringify(event),
      });
      const result = await rawResult.json();
      if (rawResult.status !== 201) {
        const details = result?.details?.body?.map(({ message }) => message);
        setError({ message: result.message, details });
        setSuccess(false);
      } else {
        setError(null);
        setSuccess(true);
      }
      setLoading(false);
    } catch {
      setLoading(false);
      setSuccess(false);
      setError({ message: 'An error occurred trying to create your event. Please try again later.' });
    }
  };

  const onBack = () => {
    setError(null);
    setSuccess(false);
  }

  return (
    <Paper className={container} data-testid="EventsFormContainer">
      <Typography variant="h2" align="center">
        Create event
      </Typography>
      {(success || error) && (
        <EventsFormStatus
          success={success}
          error={error}
          onBack={onBack}
        />
      )}
      {!success && !error && loading && (
        <Box display="flex" justifyContent="center">
          <CircularProgress size={100} className={loadIndicator} />
        </Box>
      )}
      {!success && !error && !loading && <EventsForm onSubmit={onSubmit} />}
    </Paper>
  );
};

export default EventsFormContainer;
