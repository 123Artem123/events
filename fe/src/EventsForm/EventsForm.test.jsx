import { render, screen } from "@testing-library/react";
import { MuiPickersUtilsProvider } from "@material-ui/pickers";
import MomentUtils from "@date-io/moment";
import EventsForm from "./EventsForm";
import userEvent from "@testing-library/user-event";

describe("EventsForm", () => {
  test("renders", () => {
    const { asFragment } = render(
      <MuiPickersUtilsProvider utils={MomentUtils}>
        <EventsForm onSubmit={() => {}} />
      </MuiPickersUtilsProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("should call submit", async () => {
    const onSubmit = jest.fn();
    render(
      <MuiPickersUtilsProvider utils={MomentUtils}>
        <EventsForm onSubmit={onSubmit} />
      </MuiPickersUtilsProvider>
    );
    const emailField = screen.getByTestId("EventsForm-email");
    const fnField = screen.getByTestId("EventsForm-fn");
    const lnField = screen.getByTestId("EventsForm-ln");
    const submitBtn = screen.getByTestId("EventsForm-submit");
    userEvent.type(emailField, "test@test.com");
    userEvent.type(fnField, "fn");
    userEvent.type(lnField, "ln");
    userEvent.click(submitBtn);
    expect(onSubmit).toHaveBeenCalledWith({
      date: new Date().toISOString().substring(0, 10),
      email: "test@test.com",
      firstName: "fn",
      lastName: "ln",
    });
  });
});
