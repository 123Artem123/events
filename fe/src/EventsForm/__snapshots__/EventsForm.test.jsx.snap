// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventsForm renders 1`] = `
<DocumentFragment>
  <form
    autocomplete="off"
    novalidate=""
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-justify-content-xs-center"
    >
      <div
        class="MuiFormControl-root MuiTextField-root makeStyles-input-1"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink="false"
        >
          Email
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
        >
          <input
            class="MuiInputBase-input MuiOutlinedInput-input"
            data-testid="EventsForm-email"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-2 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-4"
            >
              <span>
                Email
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root makeStyles-input-1"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink="false"
        >
          First Name
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
        >
          <input
            class="MuiInputBase-input MuiOutlinedInput-input"
            data-testid="EventsForm-fn"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-2 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-4"
            >
              <span>
                First Name
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root makeStyles-input-1"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink="false"
        >
          Last Name
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
        >
          <input
            class="MuiInputBase-input MuiOutlinedInput-input"
            data-testid="EventsForm-ln"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-2 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-4"
            >
              <span>
                Last Name
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root makeStyles-input-1"
        format="MMMM Do"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled"
          data-shrink="true"
        >
          Date
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input"
            readonly=""
            type="text"
            value="September 4th"
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-2 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-4 PrivateNotchedOutline-legendNotched-5"
            >
              <span>
                Date
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-direction-xs-column MuiGrid-justify-content-xs-space-evenly"
      >
        <div
          class="MuiGrid-root MuiGrid-item"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-outlinedSizeLarge MuiButton-sizeLarge MuiButton-fullWidth"
            tabindex="0"
            type="reset"
          >
            <span
              class="MuiButton-label"
            >
              Clear
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-outlinedSizeLarge MuiButton-sizeLarge MuiButton-fullWidth"
            data-testid="EventsForm-submit"
            tabindex="0"
            type="submit"
          >
            <span
              class="MuiButton-label"
            >
              Submit
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
